### DISCUSSION TITLE
Error tests e2e despliegue (github actions)

### DISCUSSION THREAD

**Discussion by Santiago21112001** [Arquisoft/faq#66] [Q&A] [answered]
Buenas, somos el equipo wiq_es04d. Hemos añadido tests e2e y hemos comprobado que pasan en local.
El problema es que al hacer una release, los tests e2e en el github actions fallan, y se nos hace extraño porque son los mismos.
Adjunto captura (véase [1] para más detalles):
![imagen](https://github.com/Arquisoft/faq/assets/79706368/4cb8f95a-a521-42bf-aa2d-01233acb88d3)
El resto de errores siguen el mismo formato, es decir, fallan en el await page
Esta es la configuración que usamos ahora en los tests (véase [2]):
![imagen](https://github.com/Arquisoft/faq/assets/79706368/d134e76c-21c5-411f-a06b-d1c42f5c3662)

Hemos intentado varias cosas para solucionarlo: aumentar el timeout, poner headless: "new", agregar slowMo, etc.
También hemos investigado por Internet y hemos encontrado un issue de puppeteer que habla de un problema similar (véase [3])

Ahora nos pasan porque hemos comentado el contenido de los tests que fallan.

¿Alguna idea de qué puede pasar? Gracias de antemano.

Edit: los tests para probar están en la rama **develop**

Referencias:
[1] error detallado - https://github.com/Arquisoft/wiq_es04d/actions/runs/8823749977/job/24224952179
[2] jugar-form.steps.js - https://github.com/Arquisoft/wiq_es04d/blob/master/webapp/e2e/steps/jugar-form.steps.js
[3] issue de puppeteer - https://github.com/puppeteer/puppeteer/issues/8166

**Comment by pglez82** [comment]
He clonado vuestro repositorio y ejecutado vuestros tests e2e (después de descomentarlos), en local. Los tests empiezan con el registro del usuario, pero lo que me está pasando es que una vez registrado, la página no navega a ningún sitio. Es decir, el test mete el nombre de usuario y contraseña, da al botón, se queda ahí (veo errores en la consola). No se navega a una página que haya un botón de salir ni nada.

He probado la aplicación y me pasa lo mismo. Me trato de registrar y ahí se queda.

Es decir, lo que observo en el código que hay ahora es que falla el registro de la app. Estoy probando el código de master, no se si tengo que ir a otra rama.

**Reply by Santiago21112001** [reply]
Perdón, los tests correctos no están en master. Están en la rama develop.

**Comment by pglez82** [comment]
Me pasa lo mismo. He hecho checkout en develop, he puesto headless:false para ver lo que pasa y cuando ejecuto los tests, se rellenan los dos campos de registro, se presiona el botón y sale un mensaje de que el usuario ha sido dado de alta. Pero luego yo entiendo que en el test tendríais que loguearos porque según está ahora la web sigue en la página de registro y por tanto los tests fallan porque no encuentra los componentes que buscáis.

¿Habéis ejecutado los tests en headless:false en local para ver esto que os digo?

**Reply by Verzidee** [reply]
El mensaje que indica que el usuario ha sido dado de alta se debe a que en la ejecución anterior de los tests ya registraste el mismo usuario exitosamente. Dado que tenemos una restricción que impide que dos personas compartan el mismo nombre de usuario, después de un registro exitoso el sistema te autentica automáticamente y te redirige a la página principal. 

Para resolver esto se ha de incrementar el número al final del nombre de usuario en cada test. Aquí te dejo los pasos para facilitar la modificaciones. 

- En el archivo `history.steps.js`, cambia la línea 26 de "yago121221" a "yago121222". 
- En `jugar-form.steps.js`, modifica la línea 22 de "Zohaib1222121" a "Zohaib1222122". 
- En `register-form.steps.js`, ajusta la línea 28 de "santiago1212221" a "santiago1212222". 

Recuerda aumentar el número cada vez que ejecutes los tests para evitar el mensaje de que el usuario ya ha sido dado de alta.

**Reply by pglez82** [reply]
Pero en el test-environment-setup.js estamos creando una base de datos en memoria para los tests precisamente para evitar esto.

```
mongoserver = await MongoMemoryServer.create();
```

¿Cómo es posible que se guarden los usuarios de una ejecución de tests a la siguiente?

**Reply by Verzidee** [reply]
Hmm nosotros para ejecutar los tests e2e lo que hacemos es desplegar la maquina en local y despues ejecutamos los tests e2e, no se si esa es la manera correcta de hacerlo.

**Reply by Verzidee** [reply]
Creo que acabo de encontrar el error, el problema era que en userservice se estaba conectando a nuestra db de historial y en historyservice se conectaba a nuestra db de usuarios.

**Reply by Verzidee** [reply]
Hola de nuevo.

Tras cambiar las bases de datos.

Acabo de probar los tests e2e en local, haciendo primero un start de gatewayservice,questionservice, authservice,historyservice y userservice. Despues lo que hice fue ejecutar los tests e2e y me funcionan (Esto lo he hecho sin el despliegue en local).

![image](https://github.com/Arquisoft/faq/assets/145342346/2b4e6cf1-b824-418a-ac2f-c1e033e899d0)

Nuestro componente registrarse lo que hace es si se registro correctamente, entonces se logea automaticamente y lo redirige a home de la siguiente manera:

![image](https://github.com/Arquisoft/faq/assets/145342346/cac1f189-30c0-41cb-a9d0-f710c61f5ddf)

Pero en el release me sale este problema con los 3 tests:
![image](https://github.com/Arquisoft/faq/assets/145342346/75f99a61-a7d5-42e4-9058-889a09195403)

**Reply by pglez82** [reply]
Yo estoy prácticamente seguro que el problema no está en los test E2E. He probado en local después de hacerme pull y a mi me siguen fallando. Pero es que la aplicación también me falla.

Investigando un poco más, después de registrarse (servicio userservice), se hace una petición a login (authservice). Esta segunda a mi me falla y por tanto no se loguea después de registrarse. Por eso fallan los tests.

Después de mirar el código veo que la base de datos que usa el userservice es `mongodb://localhost:27017/historydb` y la que usa el authservice es `mongodb://localhost:27017/userdb`. Así que añadís el usuario en el registro en un lao y luego buscáis para loguearos en otro sitio. 

Probablemente te funcione en local con un usuario que exista ya en esas dos BDs pero cuando te vas a github actions (o en mi ordenador), con una base de datos vacía, no funciona.

**Reply by Verzidee** [reply]
En mi mensaje anterior, te mencioné que había ejecutado los e2e después de actualizar las bases de datos, y funcionaron.

Parece que realizaste un pull de una version anterior a esa modificacion porque, actualmente, tanto en las rama master como en develop, ese problema ya está resuelto.

Aquí está el enlace al commit donde se corrigió:
https://github.com/Arquisoft/wiq_es04d/commit/f9862ab96b5111d028dc50ab0d04b79dca218971

Por favor, verifica esto de nuevo. Yo tambien creo que no se trata de un problema con los tests E2E, sino más bien con los actions, aunque no he encontrado nada que me pueda ayudar a identificar la causa.

**Reply by pglez82** [reply]
Buenas, a mi me sigue sin funcionar la aplicación (ni el registro ni el login me van y ahora ya no es cuestión de la bbdd). La única idea que se me ocurre es que tengáis alguna variable de entorno o algo que no esté en el repositorio. Lo qu está claro es que si a mi no me funciona, en los tests:e2e en github actions tampoco va a funcionar. 

Yo lo estoy ejecutando como se haría cuando se ejecuta npm run test:e2e, es decir, voy a directorio webapp/e2e y ejecuto `node test-environment-setup.js` para levantar todo el backend. Luego voy al webapp y hago npm run build y npm run prod.

Mi recomendación es que clones el repositorio en otro directorio o en otra máquina y trates de ejecutar la app como yo lo hago, a ver si te funciona en local o no. En caso de que no te funcione podrás ver que es lo que falta en el repositorio y por lo que la app en actions no va.

Otra cosilla que haría es añadir la opción jest --runInBand al ejecutar los tests:e2e. Puede que la máquina que os asignen en github actions no sea muy potente y al ejecutar todos los tests en paralelo de problemas.

**Comment by Santiago21112001** [comment, marked as answer]
Solucionado. Faltaba definir para los tests e2e la variable de entorno SECRET_KEY que usa auth-service en la gestión de tokens con "jsonwebtoken".
![imagen](https://github.com/Arquisoft/faq/assets/79706368/69b394a5-3ae1-4141-b2d8-a2bd938967d6)

También puse jest --runInBand.
Ahora pasan los tests en github actions y todo se despliega correctamente.
